common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java

```diff
@@ -320,25 +342,25 @@ public long maxChunksBeingTransferred() {
    * threads for the completion of sending back responses, we are able to put a limit on
    * the max number of threads from TransportServer's default EventLoopGroup that are
    * going to be consumed by writing response to ChunkFetchRequest, which are I/O intensive
-   * and could take long time to process due to disk contentions. By configuring a slightly
+   * and could take long time to process due to disk contentions. By configuring a
    * higher number of shuffler server threads, we are able to reserve some threads for
    * handling other RPC messages, thus making the Client less likely to experience timeout
-   * when sending RPC messages to the shuffle server. The number of threads used for handling
-   * chunked fetch requests are percentage of io.serverThreads (if defined) else it is a percentage
-   * of 2 * #cores. However, a percentage of 0 means netty default number of threads which
-   * is 2 * #cores ignoring io.serverThreads. The percentage here is configured via
-   * spark.shuffle.server.chunkFetchHandlerThreadsPercent. The returned value is rounded off to
-   * ceiling of the nearest integer.
+   * when sending RPC messages to the shuffle server. The number of server threads needs to
+   * be a multiple of the number of threads used for handling chunked fetch requests. See
+   * SPARK-29206 for more details. This is controlled via configuring a ratio between
+   * io.serverThreads and chunk fetch handler threads, such that # server threads = ratio *
+   * # chunk fetch handler threads. The default value of this ratio is 1, which means the
+   * numbers of threads in both thread pools are equal. If io.serverThreads is not defined,
+   * then 2 * #cores threads will be used for both server threads and chunk fetch handler
+   * threads, ignoring the ratio configuration.
    */
   public int chunkFetchHandlerThreads() {
     if (!this.getModuleName().equalsIgnoreCase("shuffle")) {
       return 0;
     }
-    int chunkFetchHandlerThreadsPercent =
-      conf.getInt("spark.shuffle.server.chunkFetchHandlerThreadsPercent", 100);
-    int threads =
-      this.serverThreads() > 0 ? this.serverThreads() : 2 * NettyRuntime.availableProcessors();
-    return (int) Math.ceil(threads * (chunkFetchHandlerThreadsPercent / 100.0));
+    int chunkFetchHandlerThreadsRatio = getChunkFetchHandlerThreadsRatio();
+    return this.serverThreads() > 0 ? this.serverThreads() / chunkFetchHandlerThreadsRatio :
```

Comment:
ok this makes a bit more sense but it doesn't work with the above logic..
9/2 = 4  and server threads ends up with same value, you need one or the other.

Reply:
It won't be 9/2, it would be 10/2.
this.serverThreads will change the configured 9 server threads to 10 to make sure it can be divided by the int ratio.

Reply:
In the calculation of `chunkFetchHandlerThreads()` the `chunkFetchHandlerThreadsRatio` is applied twice (once within the  serverThreads def) which gives quite a few choices and with quite low values for `chunkFetchHandlerThreads`. For example let's see the case where the availableProcessors is 16 and depending on the value of SPARK_NETWORK_IO_SERVERTHREADS_KEY we got the followings:

SPARK_NETWORK_IO_SERVERTHREADS_KEY = 0

chunkFetchHandlerThreadsRatio |  serverThreads() | chunkFetchHandlerThreads()
------|-------------|----------------| 
1   |     0       |   32           |
2   |     0       |   32           |
3   |     0       |   32           |
4   |     0       |   32           |
5   |     0       |   32           |

*So in this case ratio ha no effect at all.* The `chunkFetchHandlerThreads` is always the double of the available processors.

SPARK_NETWORK_IO_SERVERTHREADS_KEY = 4 

chunkFetchHandlerThreadsRatio |  serverThreads() | chunkFetchHandlerThreads()
------|-------------|----------------| 
1   |     4       |   4           |
2   |     2       |   1           |
3   |     2       |   0           |
4   |     1       |   0           |
5   |     1       |   0           |


SPARK_NETWORK_IO_SERVERTHREADS_KEY = 8

chunkFetchHandlerThreadsRatio |  serverThreads() | chunkFetchHandlerThreads()
------|-------------|----------------| 
1   |     8       |   8           |
2   |     4       |   2           |
3   |     3       |   1           |
4   |     2       |   0           |
5   |     2       |   0           |


## Before this change 

For comparisation here are the values for the different `chunkFetchHandlerThreadsPercent`:

SPARK_NETWORK_IO_SERVERTHREADS_KEY = 0

chunkFetchHandlerThreadsPercent |  serverThreads() | chunkFetchHandlerThreads()
------|-------------|----------------| 
25%   |     0       |   8            |
50%   |     0       |   16           |
75%   |     0       |   24           |
100%  |     0       |   32           |


SPARK_NETWORK_IO_SERVERTHREADS_KEY = 4

chunkFetchHandlerThreadsPercent |  serverThreads() | chunkFetchHandlerThreads()
------|-------------|----------------| 
25%   |     4       |   1           |
50%   |     4       |   2           |
75%   |     4       |   3           |
100%  |     4       |   4           |

SPARK_NETWORK_IO_SERVERTHREADS_KEY = 8

chunkFetchHandlerThreadsPercent |  serverThreads() | chunkFetchHandlerThreads()
------|-------------|----------------| 
25%   |     8       |   2           |
50%   |     8       |   4           |
75%   |     8       |   6           |
100%  |     8       |   8           |

Reply:
The other problem I see with this when `chunkFetchHandlerThreads` reaches 0 then it is handled as 2x#cores bu Netty which much more than `serverThreads` for those cases. 

What about setting `chunkFetchHandlerThreads` to 1 in that case?